### PR TITLE
VStreamer: For larger compressed transaction payloads, stream the internal contents

### PR DESCRIPTION
**Pull request by mattlord**
## Description

For larger compressed transaction payloads (> ZstdInMemoryDecompressorMaxSize) we were already streaming the internal events as we decompressed the payload, but in the vstreamer we were still reading the entire contents into memory before sending them to the consumer (vplayer).

In this PR, we stream the internal contents all the way from the binlog consumer to the vstream consumer so that we do not need to hold the entire contents, which can be 10s or even 100s of GiBs, in memory all at once.  As you can see in the test/demonstration below, we allocate and use DRASTICALLY less memory when processing the payloads: in this case **_approximately 14-18 times less_**.

Here's a manual test/demonstration on macOS (note that we end up with over 40 million rows in the customer table):
```
make build 

cd examples/local
alias vtctldclient='command vtctldclient --server=localhost:15999'

# Setup function to wait for the copy phase to complete
function wait_for_workflow_running() {
    local keyspace=customer
    local workflow=commerce2customer
    local wait_secs=900
    local result=""

    echo "Waiting for the ${workflow} workflow in the ${keyspace} keyspace to finish the copy phase..."

    for _ in $(seq 1 ${wait_secs}); do
        result=$(vtctldclient Workflow --keyspace="${keyspace}" show --workflow="${workflow}" 2>/dev/null | grep "Copy phase completed")
        if [[ ${result} != "" ]]; then
            break
        fi
        sleep 1
    done;

    if [[ ${result} == "" ]]; then
        echo "Timed out after ${wait_secs} seconds waiting for the ${workflow} workflow in the ${keyspace} keyspace to reach the running state"
    else
        echo "The ${workflow} workflow in the ${keyspace} keyspace is now running. $(sed -rn 's/.*"(Copy phase.*)".*/\1/p' <<< "${result}")."
    fi
}


./101_initial_cluster.sh; ./201_customer_tablets.sh


# Enable binlog transaction compression
for uid in $(vtctldclient GetTablets | awk '{print $1}' | cut -d- -f2 | bc); do
  command mysql -u root --socket "${VTDATAROOT}/vt_0000000${uid}/mysql.sock" -e "stop replica; set global  binlog_transaction_compression=ON; start replica;"
done


# Load data in the customer table
commerce_primary_uid=$(vtctldclient GetTablets --keyspace commerce --tablet-type primary --shard "0" | awk '{print $1}' | cut -d- -f2 | bc)
table_file="${VTDATAROOT}/vt_0000000${commerce_primary_uid}/data/vt_commerce/customer.ibd"

# Generate 5MiB of initial data
size=$((5*1024*1024))
while [[ $(stat -f "%z" "${table_file}") -lt ${size} ]]; do
    command mysql -u root --socket "${VTDATAROOT}/vt_0000000${commerce_primary_uid}/mysql.sock" vt_commerce -e "insert into customer (customer_id, email) values (${RANDOM}*${RANDOM}, '${RANDOM}_person@planetscale.com')" 2> /dev/null
done

say "Initial data load completed"

# Grow that to at least 2GiB
size=$((2*1024*1024*1024))
i=1
while [[ $(stat -f "%z" "${table_file}") -lt ${size} ]]; do
    command mysql -u root --socket "${VTDATAROOT}/vt_0000000${commerce_primary_uid}/mysql.sock" vt_commerce -e "insert into customer (email) select concat(${i}, email) from customer limit 5000000"
    let i=i+1
done

say "Full data load completed"

# Move the customer table from commerce to customer
vtctldclient MoveTables --workflow commerce2customer --target-keyspace customer create --source-keyspace commerce --tables "customer,corder" --tablet-types="primary"
wait_for_workflow_running

say "Workflow is running"

rm /tmp/vttablet-sample

sample $(cat ${VTDATAROOT}/vt_0000000${commerce_primary_uid}/vttablet.pid) 600 -file /tmp/vttablet-sample &

command mysql -u root --socket "${VTDATAROOT}/vt_0000000${commerce_primary_uid}/mysql.sock" vt_commerce -e "update customer set email = concat('update1_', email) limit 10000000"

say "Bulk update complete"


# Wait for the transaction payload event to be processed
while [[ $(curl -s "http://localhost:15${commerce_primary_uid}/debug/vars" | jq -r '.CompressedTransactionPayloadsViaStream') -ne 1 ]]; do
    sleep 1
done
# Wait for the workflow to catch up
sleep 5
while [[ $(vtctldclient MoveTables --target-keyspace=customer --workflow=commerce2customer show --compact --include-logs=false | jq -r '.workflows[0].max_v_replication_transaction_lag') -gt 1 ]]; do
    sleep 1
done
say "Bulk update has been successfully replicated"

kill -INT %1
grep "Physical footprint" /tmp/vttablet-sample
rm /tmp/vttablet-sample

say "Update test complete"


sample $(cat ${VTDATAROOT}/vt_0000000${commerce_primary_uid}/vttablet.pid) 1200 -file /tmp/vttablet-sample &

command mysql -u root --socket "${VTDATAROOT}/vt_0000000${commerce_primary_uid}/mysql.sock" vt_commerce -e "update customer set email = concat('update2_', email) limit 20000000"

say "Bulk update two complete"


# Wait for the transaction payload event to be processed
while [[ $(curl -s "http://localhost:15${commerce_primary_uid}/debug/vars" | jq -r '.CompressedTransactionPayloadsViaStream') -ne 2 ]]; do
    sleep 1
done
# Wait for the workflow to catch up
sleep 5
while [[ $(vtctldclient MoveTables --target-keyspace=customer --workflow=commerce2customer show --compact --include-logs=false | jq -r '.workflows[0].max_v_replication_transaction_lag')  -gt 1 ]]; do
    sleep 1
done
say "Bulk update has been successfully replicated"

kill -INT %1
grep "Physical footprint" /tmp/vttablet-sample
rm /tmp/vttablet-sample

say "Update test two complete"


# Do a vdiff to be sure everything was replicated correctly
vtctldclient vdiff --target-keyspace customer --workflow commerce2customer create --tablet-types=primary --wait
```

### Results on the PR branch:
```
update1:
Physical footprint:         310.5M
Physical footprint (peak):  310.5M

update2:
Physical footprint:         586.5M
Physical footprint (peak):  747.3M
```

### Results on the main branch:
```
update1:
Physical footprint:         5.1G
Physical footprint (peak):  5.7G

update2:
Physical footprint:         10.3G
Physical footprint (peak):  10.9G
```

## Related Issue(s)
 
  - Follow-up to
    - https://github.com/vitessio/vitess/pull/16328
    - https://github.com/vitessio/vitess/pull/17220

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by rohit-nayak-ps — APPROVED** [review]
Changes lgtm.

Are we testing binlog compression in unit tests or e2e atm?

**Comment by mattlord** [comment]
> Changes lgtm.
> 
> Are we testing binlog compression in unit tests or e2e atm?

We have the updated unit test for the lower level part.

And it's enabled in these e2e tests:
```
.github/workflows/cluster_endtoend_onlineddl_vrepl.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_onlineddl_vrepl_stress.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_onlineddl_vrepl_stress_suite.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_onlineddl_vrepl_suite.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_schemadiff_vrepl.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_across_db_versions.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_basic.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_cellalias.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_copy_parallel.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_foreign_key_stress.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_mariadb_to_mysql.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_migrate.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_multi_tenant.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_partial_movetables_and_materialize.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_v2.yml:        binlog-transaction-compression=ON
.github/workflows/cluster_endtoend_vreplication_vtctldclient_vdiff2_movetables_tz.yml:        binlog-transaction-compression=ON
test/templates/cluster_endtoend_test.tpl:        binlog-transaction-compression=ON
```

File: go/mysql/binlog_event_compression.go [hunk 1]
```diff
@@ -292,6 +297,8 @@ func (tp *TransactionPayload) decompress() error {
 		}
 		compressedTrxPayloadsUsingStream.Add(1)
 		tp.reader = streamDecoder
+		// Signal the consumer to also stream the contents.
+		tp.StreamingContents = true
```
Comment (shlomi-noach) [inline go/mysql/binlog_event_compression.go:301]: Should we just always stream? (Meaning we would remove the conditional flag?)

I guess there's also a backwards-compatibility issue here?

**Review by shlomi-noach — APPROVED** [review]

File: go/mysql/binlog_event_compression.go (hunk 1 above)
Comment (mattlord) [reply go/mysql/binlog_event_compression.go:301]: I thought about that too. It's more efficient to do it in a batch though so for smaller payloads we do that (one batch) at both layers, binlog and vstreamer.

**Pull request merged**